### PR TITLE
Fix branch filtering and source encoding

### LINKED ISSUE #14: Error with non-utf8 source files
```
Exception in thread Thread-2:
Traceback (most recent call last):
  File "/usr/lib/python3.7/threading.py", line 917, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.7/threading.py", line 865, in run
    self._target(*self._args, **self._kwargs)
  File "/home/daniel/Vcs/fastcov/fastcov.py", line 196, in exclMarkerWorker
    for i, line in enumerate(f, 1): #Start enumeration at line 1
  File "/usr/lib/python3.7/codecs.py", line 322, in decode
    (result, consumed) = self._buffer_decode(data, self.errors, final)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xa1 in position 8135: invalid start byte
```

The file in question: https://github.com/vim/vim/blob/94f82cbacf76767b5ac32f813e1d670501dbd0e6/src/digraph.c

### LINKED ISSUE #15: Branches destroyed due to --exclude-br-lines-starting-with
`--exclude-br-lines-starting-with`, if absent (but with branch coverage enabled), causes nearly all branches to be filtered. 

This is due to an implementation bug.

### PR DESCRIPTION
**Pull request by RPGillespie6**
Description:
- Fix #14
- Fix #15
- Add new option `--fallback-encodings` that allows specifying fallback codecs fastcov should try
- Bump version to 1.2

### CODE REVIEW

File: fastcov.py [hunk 1]
```diff
@@ -184,55 +184,68 @@ def dumpToLcovInfo(fastcov_json, output):
...
-            for i, line in enumerate(f, 1): #Start enumeration at line 1
-                if str(i) in fastcov_sources[source]["branches"]:
-                    if all(not line.lstrip().startswith(e) for e in include_branches_sw): # Include branches starting with...
-                        del fastcov_sources[source]["branches"][str(i)]
+        for i, line in enumerate(getSourceLines(source, fallback_encodings), 1): #Start enumeration at line 1
+            if str(i) in fastcov_sources[source]["branches"]:
+                if include_branches_sw and all(not line.lstrip().startswith(e) for e in include_branches_sw): # Include branches starting with...
+                    del fastcov_sources[source]["branches"][str(i)]
+
+                if exclude_branches_sw and any(line.lstrip().startswith(e) for e in exclude_branches_sw): # Exclude branches starting with...
```
Comment (blueyed) [inline fastcov.py]: `exclude_branches_sw and` not needed here, but more explicit.

Comment (RPGillespie6) [reply fastcov.py]: Yes, I made them both explicit since I was burned previously by the implicit behavior of `all([])` vs `any([])` :P

**Comment by blueyed** [comment]
Sorry for hijacking this PR / asking here, but I've wondered if you have experience with using it on codecov?
I've tried this branch (before it was deleted ;)), but coverage dropped massively: https://codecov.io/gh/neovim/neovim/tree/3535fb141436131cc7362b3dc46f37ffe28984da (should be rather >70%).  Any ideas?
PR: https://github.com/neovim/neovim/pull/10603/files#diff-cbfef1436fe72c07592f5f2a83ca5ffc
This was using `--lcov`, trying `--gcov-raw` now.

**Comment by RPGillespie6** [comment]
I've never used codecov, no, but I can check it out. What intermediate formats does codecov support? 

Is the coverage correct if the report is generated with genhtml instead of codecov?

```
$ fastcov -b --lcov -o lcov.info --gcov "$GCOV" 
$ genhtml -o code_coverage report.info
$ firefox code_coverage/index.html
```

From the link you posted it looks like previously it was pushing an `xml` file (and fastcov currently only generates lcov and json-based formats)

**Comment by RPGillespie6** [comment]
Hmm, according to [this](https://docs.codecov.io/reference#section-acceptable-report-formats) codecov should accept lcov format. 

Let me try cloning and building neovim

**Comment by blueyed** [comment]
Yes, genhtml locally looks OK.  Therefore it might not help/make a difference if you try it locally yourself.
I've used `coverage.xml` before with gcovr, because it does not support other formats suitable for codecov.

**Comment by RPGillespie6** [comment]
Hmm, could you post your lcov info file somewhere so I could take a look? I want to examine the paths of the source files in it. I suspect that because fastcov converts all paths to be absolute, codecov might not be able to "find" the source files after it is uploaded if it doesn't have the same directory structure.

If not, I can open a ticket with codecov

**Comment by blueyed** [comment]
The uploaded lcov.info files can be seen/downloaded via the "Builds" tab: https://codecov.io/gh/neovim/neovim/commit/0efc415f2b44b057534084896dc3ca54517977aa/build
The file paths therein look OK to me, i.e. they are all absolute.

It might be that codecov is not merging the reports here properly?! (we have 3 uploads with different flags).
I am trying to only upload for "functionaltest" now.

**Comment by RPGillespie6** [comment]
So codecov is sharing the same VM/container as Travis, then? (i.e. codecov can see `/home/travis/build/`)? I thought I saw that codegen maintains its own clone based on a github url. We could look in the gcovr XML file to see what the paths look like in that (relative vs. absolute).

As a small experiment, you could do the following to the lcov info files before uploading and see if it makes a difference: 

`sed -i 's/\/home\/travis\/build\/neovim\/neovim/\./g' report.info`

If worse comes to worse, we can always add support for another intermediate format that is known to work with codecov, but I'd also like to open an issue with codecov if it indeed has a report merging issue.

**Comment by RPGillespie6** [comment]
It looks like the paths in the XML are relative to the neovim repo root, but also seem to reference the travis build: `filename="build/../src/nvim/memory.c"` so perhaps you are right.

However, I found this: https://docs.codecov.io/docs/fixing-paths

> For Codecov to operate properly, all files must match the [git] file structure

I'm curious if the `sed` command which makes the paths relative to the git root fixes it

**Comment by blueyed** [comment]
According to https://github.com/codecov/example-cpp11-cmake#travis-setup-using-lcov the lcov format should be supported in general.

Using `lcov` on the output gives a warning btw:

> lcov: WARNING: negative counts found in tracefile build/lcov.info

`sed` does not help - with the following result:
```
<<<<<< network
# path=lcov.info
SF:src/nvim/memory.c
```

**Comment by blueyed** [comment]
Using `lcov` to "massage" the file appears to fix it in general, but it is missing branch coverage there then:
```
+lcov --remove lcov.info /usr/* --output-file coverage.info
Reading tracefile lcov.info
Deleted 0 files
Writing data to coverage.info
Summary coverage rate:
  lines......: 69.5% (83332 of 119912 lines)
  functions..: 78.2% (4351 of 5564 functions)
  branches...: no data found
```

**Comment by blueyed** [comment]
When using `lcov --rc lcov_branch_coverage=1` with some `--remove` it looks much better in codecov (i.e. the massaging appears to fix things).

btw: according to http://ltp.sourceforge.net/coverage/lcov/geninfo.1.php the FN entry is meant to be absolute.

**Comment by blueyed** [comment]
Not sure what this means, but I am seeing negative branch counts?!
```
(Pdb++) branch
{'fallthrough': False, 'count': -1, 'throw': False}
(Pdb++) line_raw
{'branches': [{'fallthrough': True, 'count': 306924, 'throw': False}, {'fallthrough': False, 'count': -1, 'throw': False}], 'count': 306923, 'line_number': 149, 'unexecuted_block': False, 'function_name': 'multiqueue_process_events'}
(Pdb++) line_number
'149'
```

I've looked a bit into this, and started fixing things in https://github.com/RPGillespie6/fastcov/pull/18.

**Pull request merged**